source/bitbucket/bitbucket.go

```diff
@@ -0,0 +1,213 @@
+package bitbucket
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	nurl "net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang-migrate/migrate/v4/source"
+	"github.com/ktrysmt/go-bitbucket"
+)
+
+func init() {
+	source.Register("bitbucket", &Bitbucket{})
+}
+
+var (
+	ErrNoUserInfo             = fmt.Errorf("no username:password provided")
+	ErrNoAccessToken          = fmt.Errorf("no password/app password")
+	ErrInvalidRepo            = fmt.Errorf("invalid repo")
+	ErrInvalidBitbucketClient = fmt.Errorf("expected *bitbucket.Client")
+	ErrNoDir                  = fmt.Errorf("no directory")
+)
+
+type Bitbucket struct {
+	config     *Config
+	client     *bitbucket.Client
+	migrations *source.Migrations
+}
+
+type Config struct {
+	Owner string
+	Repo  string
+	Path  string
+	Ref   string
+}
+
+func (b *Bitbucket) Open(url string) (source.Driver, error) {
+	u, err := nurl.Parse(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.User == nil {
+		return nil, ErrNoUserInfo
+	}
+
+	password, ok := u.User.Password()
+	if !ok {
+		return nil, ErrNoAccessToken
+	}
+
+	cl := bitbucket.NewBasicAuth(u.User.Username(), password)
+
+	bi := &Bitbucket{
+		client:     cl,
+		migrations: source.NewMigrations(),
+	}
+
+	bi.ensureFields()
+
+	// set owner, repo and path in repo
+	bi.config.Owner = u.Host
+	pe := strings.Split(strings.Trim(u.Path, "/"), "/")
+	if len(pe) < 1 {
+		return nil, ErrInvalidRepo
+	}
+	bi.config.Repo = pe[0]
+	if len(pe) > 1 {
+		bi.config.Path = strings.Join(pe[1:], "/")
+	}
+	bi.config.Ref = u.Fragment
+
+	if err := bi.readDirectory(); err != nil {
```

Comment:
Call `WithInstance()` so there's a single path to instantiate a `Bitbucket` struct.

Reply:
done!!